lib/redis_client.rb

```diff
@@ -692,6 +692,22 @@ def connect
       raise
     end
   end
+
+  def with_middlewares(method, *args, &block)
+    Middlewares.send(method, *args) do
+      call_local_middlewares(config.middlewares, 0, method, *args, &block)
+    end
+  end
+
+  def call_local_middlewares(middlewares, position, method, *args, &block)
+    middleware = middlewares[position]
+
+    return block.call if middleware.nil?
+
+    middleware.send(method, *args) do
```

Comment:
Similarly, using `send` is significantly slower because it doesn't benefit from method caches.

Reply:
Yes:

```ruby
>> puts RubyVM::InstructionSequence.compile("1.bar").disasm
== disasm: #<ISeq:<compiled>@<compiled>:1 (1,0)-(1,5)> (catch: FALSE)
0000 putobject_INT2FIX_1_                                             (   1)[Li]                           
0001 opt_send_without_block                 <calldata!mid:bar, argc:0, ARGS_SIMPLE>                        
0003 leave                                                                                                 
=> nil                                                                                                     
>> puts RubyVM::InstructionSequence.compile("1.send(:bar)").disasm
== disasm: #<ISeq:<compiled>@<compiled>:1 (1,0)-(1,12)> (catch: FALSE)
0000 putobject_INT2FIX_1_                                             (   1)[Li]                           
0001 putobject                              :bar                                                         
0003 opt_send_without_block                 <calldata!mid:send, argc:1, ARGS_SIMPLE>                     
0005 leave   
```

When you do `send(:foo)`, there is a cache for looking up the `send` method, none of the `foo` method, the same issue applies to `respond_to?`.

That said, it's what is is today, doesn't necessarily apply to other implementations and might get better in the future, but in general it's best to avoid `send` in hot paths.

Reply:
Thanks for the details 👍